// loi/lib/helpers/error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createError = exports.createErrorDetailItem = void 0;
const path_1 = require("../utilties/path");
const stringify_1 = require("../utilties/stringify");
const message_1 = require("./message");
/** @internal */
function createErrorDetailItem(v, context) {
    return {
        type: "invalid",
        path: path_1.getRealContextPath(context),
        kind: context[context.length - 1].type.name,
        message: `Invalid value \`${stringify_1.stringify(v)}' supplied to ${path_1.getRealContextPath(context)}: ${context[context.length - 1].type.name}.`
    };
}
exports.createErrorDetailItem = createErrorDetailItem;
function createError(results) {
    const error = new Error();
    error.httpCode = 400;
    error.code = "validation_error";
    error.message = `Validation Error\n\n${message_1.createMessage(results)}`;
    error.details = results.map((e) => createErrorDetailItem(e.value, e.context));
    return error;
}
exports.createError = createError;
